File: 02_Sorting/Quick Sort/quickSort.cpp
#include <bits/stdc++.h>
using namespace std;

int my_partition(vector<int> &arr, int low, int high){
  int pivot = arr[low];
  int i = low;
  int j = high;

  while( i < j ){
    while( arr[i] <= pivot && i <= high - 1){
      ++i;
    }

    while( arr[j] > pivot && j >= low + 1){
      --j;
    }

    if(i < j){
      swap(arr[i], arr[j]);
    }
  }

  swap(arr[low], arr[j]);

  return j;
}

void qS(vector<int> &arr, int low, int high){
  if (low < high){
    int partition_index = my_partition(arr, low, high);
    qS(arr, low, partition_index - 1);
    qS(arr, partition_index + 1, high);
  }
}

void quickSort(vector<int> &arr){
  qS(arr, 0, arr.size()-1);
}

int main(int argc, char* argv[]){
  vector<int> v;

  int size;
  cin >> size;

  for(int i=0; i<size; ++i){
    int a;
    cin >> a;
    v.push_back(a);
  }

  quickSort(v);

  for(int i=0; i<size; ++i){
    cout << v[i] << " ";
  }
return 0;
}

File: 02_Sorting/Quick Sort/quick_sort.py
def my_partition(arr: list, low: int, high: int):
  pivot = arr[low]
  i = low
  j = high

  while i < j:
    while (arr[i] <= pivot and i <= high-1):
      i += 1

    while (arr[j] > pivot and j >= low+1):
      j -= 1

    if i < j:
      arr[i], arr[j] = arr[j], arr[i]
  
  arr[low], arr[j] = arr[j], arr[low]

  return j
  
def qS(arr: list, low: int, high: int):
  if low < high:
    partition_index = my_partition(arr, low, high)
    qS(arr, low, partition_index-1)
    qS(arr, partition_index+1, high)

def quick_sort(arr: list):
  low = 0
  high = len(arr)-1
  qS(arr, low, high)

size = int(input("Enter size: "))
arr = list(map(int, input("Enter elements in array: ").strip().split()[:size]))

quick_sort(arr)

print(arr)